src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.Object.cs

```diff
@@ -311,6 +311,176 @@ public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOp
+
+                if (reader.TokenType == JsonTokenType.False)
+                {
+                    return false;
+                }
+
+                if (reader.TokenType == JsonTokenType.Number)
+                {
+                    if (reader.TryGetInt32(out int i))
+                    {
+                        return i;
+                    }
+
+                    return reader.GetDouble();
+                }
+
+                if (reader.TokenType == JsonTokenType.String)
+                {
+                    if (reader.TryGetDateTime(out DateTime datetime))
+                    {
+                        return datetime;
+                    }
+
+                    return reader.GetString();
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+            {
+                WriteCallCount++;
+
+                if (value is int i)
+                {
+                    writer.WriteNumberValue(i);
+                }
+                else if (value is bool b)
+                {
+                    writer.WriteBooleanValue(b);
+                }
+                else if (value is string s)
+                {
+                    writer.WriteStringValue(s);
+                }
+                else
+                {
+                    throw new NotSupportedException();
+                }
+            }
+        }
+
+        private class ClassWithPrimitives
+        {
+            public int MyIntProperty { get; set; }
+            public bool MyBoolProperty { get; set; }
+            public string MyStringProperty { get; set; }
+#pragma warning disable 0649
+            public int MyIntField;
+            public bool MyBoolField;
+            public string MyStringField;
+#pragma warning restore
+        }
+
+        [Fact]
+        public static void ClassWithPrimitivesObjectConverterDeserialize()
+        {
+            const string json =
+                @"{""MyIntProperty"":123,""MyBoolProperty"":true,""MyStringProperty"":""Hello""," +
+                @"""MyIntField"":321,""MyBoolField"":true,""MyStringField"":""World""}";
+
+            var converter = new PrimitiveConverter();
+            var options = new JsonSerializerOptions
+            {
+                IncludeFields = true
+            };
+            options.Converters.Add(converter);
+
+            var obj = JsonSerializer.Deserialize<ClassWithPrimitives>(json, options);
+
+            Assert.Equal(6, converter.ReadCallCount);
+
+            Assert.Equal(123, obj.MyIntProperty);
+            Assert.True(obj.MyBoolProperty);
+            Assert.Equal("Hello", obj.MyStringProperty);
+            Assert.Equal(321, obj.MyIntField);
+            Assert.True(obj.MyBoolField);
+            Assert.Equal("World", obj.MyStringField);
+        }
+
+        [Fact]
+        public static void ClassWithPrimitivesObjectConverterSerialize()
```

Comment:
We try to combine the serialize and deserialize tests (in this case `ClassWithPrimitivesObjectConverterDeserialize` and `ClassWithPrimitivesObjectConverterSerialize`) for each scenario, so that they are "roundtrip tests". Please combine here and elsewhere.